src/main/java/org/hydev/mcpm/client/installer/InstallInteractor.java

```diff
@@ -57,38 +55,31 @@ public void installPlugin(InstallInput input) throws InstallException
         }
 
         var pluginVersion = latestPluginModel.getLatestPluginVersion()
-            .orElseThrow(() -> new InstallException(Type.NO_VERSION_AVAILABLE));
+                .orElseThrow(() -> new InstallException(Type.NO_VERSION_AVAILABLE));
+        databaseManager.checkPluginInstalled(pluginVersion);
```

Comment:
I see you took out the code that used to be here and put it into a method

But there's still a problem that if a dependency (including the initial plugin) is already installed, the rest of the installation will abort because of the thrown exception. 
In terms of the expected behaviour for the use case, we would want it to skip over these plugins, not to stop the installation.
Taylor suggested that we can return the state of the entire installation. The states could be like this: success and fail. Success is if EVERY installation either succeeds or is skipped because it is already installed (NOT a fail). A fail is if there is an error somehow, for AT LEAST one of the plugins.

You could do something like this (here I will use true and false for success and fail, but you should wrap it with a Record):
```suggestion
        if (searchResult.plugins().isEmpty()) {
             // throw new InstallException(Type.NOT_FOUND);
             return false; // we failed so return false
        }
        
...

       var name = pluginVersion.meta().name();
         List<PluginYml> pluginInstalled = input.pluginTracker().listInstalled();
         for (PluginYml pluginYml : pluginInstalled) {
             if (pluginYml != null && pluginYml.name() != null && pluginYml.name().equals(name)) {
                 return true; // we skip the installed plugin so return true because we did not fail
             }
         }

...

        // 4. Installing the depency of that plugin
         if (pluginVersion.meta().depend()!= null) {
             boolean ret = true;
             for (String dependency:pluginVersion.meta().depend()) {
                 InstallInput dependencyInput = new InstallInput(dependency,
                         SearchPackagesType.BY_NAME,
                         installInput.filePath(),
                         installInput.load());
                 ret &= installPlugin(dependencyInput); // we return true iff ALL installs succeed (do not fail), so we use and
 //                input.pluginTracker().removeManuallyInstalled(name);
             }
         }
```

Reply:
Obviously you can't do &= with a record. But what you can do is:
```suggestion
public record InstallResult
{
    State state, // this corresponds to whether failedPlugins is empty or not
    Collection<PluginModel> failedPlugins;
    public enum State {
        SUCCESS,
        FAILED
    }
    public void merge(InstallResult other) {
        if (this.state == State.SUCCESS && other.state == State.SUCCESS) {
            return;
        }
        state.State = State.FAILED;
        failedPlugins.addAll(other.failedPlugins);
}
```
and then in the install method:
```suggestion
        // 4. Installing the depency of that plugin
         InstallResult ret = new InstallResult(InstallResult.State.SUCCESS, List.of());
         if (pluginVersion.meta().depend()!= null) {
             for (String dependency:pluginVersion.meta().depend()) {
                 InstallInput dependencyInput = new InstallInput(dependency,
                         SearchPackagesType.BY_NAME,
                         installInput.filePath(),
                         installInput.load());
                 ret.merge(installPlugin(dependencyInput)); // we return true iff ALL installs succeed (do not fail), so we use and
 //                input.pluginTracker().removeManuallyInstalled(name);
             }
         }
         return ret;
```